src/olympia/lib/crypto/signing.py

```diff
@@ -0,0 +1,282 @@
+import hashlib
+import os
+import shutil
+import tempfile
+import zipfile
+
+from base64 import b64decode, b64encode
+
+from django.conf import settings
+from django.core.files.storage import default_storage as storage
+from django.utils.encoding import force_bytes
+
+import requests
+import six
+import waffle
+
+from django_statsd.clients import statsd
+from requests_hawk import HawkAuth
+from signing_clients.apps import JarExtractor
+from asn1crypto import cms
+
+import olympia.core.logger
+
+from olympia import amo
+
+
+log = olympia.core.logger.getLogger('z.crypto')
+
+
+SIGN_FOR_APPS = (amo.FIREFOX.id, amo.ANDROID.id)
+
+
+class SigningError(Exception):
+    pass
+
+
+def supports_firefox(file_obj):
+    """Return True if the file supports Firefox or Firefox for Android.
+
+    We only sign files that are at least compatible with Firefox/Firefox for
+    Android.
+    """
+    apps = file_obj.version.apps.all()
+    return apps.filter(max__application__in=SIGN_FOR_APPS)
+
+
+def get_id(addon):
+    """Return the addon GUID if <= 64 chars, or its sha256 hash otherwise.
+
+    We don't want GUIDs longer than 64 chars: bug 1203365.
+    """
+    guid = addon.guid
+    if len(guid) <= 64:
+        return guid
+    return hashlib.sha256(guid).hexdigest()
+
+
+def call_signing(file_obj):
+    """Get the jar signature and send it to the signing server to be signed."""
+    # Extract jar signature.
+    conf = settings.AUTOGRAPH_CONFIG
+    log.debug('Calling autograph service: {0}'.format(
+        conf['server_url']))
+
+    if waffle.sample_is_active('activate-autograph-file-signing'):
```

Comment:
Because of the waffle this `call_signing()` function is now super long and hard to follow. Can you break it down?

Reply:
Also, add a migration to add the default values for the waffle in the database ?

Reply:
Fixed in 4590507769, I hope that's a bit easier to read?